main.nf

```diff
@@ -16,15 +16,101 @@ nextflow.enable.dsl = 2
+        params.bwa_index = WorkflowMain.getGenomeAttribute(params, 'bwa')
+    }
+}
+
+if(!params.containsKey('regions_mask')){
+    params.regions_mask = null
+    log.info"Regions mask not specified. The entire genome will be used for alignment"
+}
 
+if(!params.containsKey('additional_fasta')){
+    params.additional_fasta = null
+    log.info"Additional fasta not specified. No additional sequences beyond those in the genome fasta will be used for alignment"
+}
+
+if (!params.containsKey('fasta_index')){
+    params.fasta_index = null
+    log.info"Genome fasta will be indexed in the workflow"
+}
+
+// these parameters are used in the pipeline and may or may not be set either
+// by the user or through the --genome argument. If they are null, then
+// the appropriate index will be created in the workflow
+if(!params.containsKey('bwa_index')){
+    params.bwa_index = null
+}
+if(!params.containsKey('bwamem2_index')){
+    params.bwamem2_index = null
+}
+if(!params.containsKey('bowtie_index')){
+    params.bowtie_index = null
+}
+if(!params.containsKey('bowtie2_index')){
+    params.bowtie2_index = null
+}
 /*
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     VALIDATE & PRINT PARAMETER SUMMARY
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 */
 
+include { validateParameters; paramsHelp } from 'plugin/nf-validation'
+
+// Print help message if needed
+if (params.help) {
+    def logo = NfcoreTemplate.logo(workflow, params.monochrome_logs)
+    def citation = '\n' + WorkflowMain.citation(workflow) + '\n'
+    def String command = "nextflow run ${workflow.manifest.name} --input samplesheet.csv --genome R64-1-1 -profile singularity"
+    log.info logo + paramsHelp(command) + citation + NfcoreTemplate.dashedLine(params.monochrome_logs)
+    System.exit(0)
+}
+
+// Validate input parameters
+if (params.validate_params) {
+    validateParameters()
+}
+
+// Validate some of the more idiosyncratic parameters specific to callingcards
+if (params.split_fastq_by_size != null && params.split_fastq_by_part != null){
+    exit 1, 'You have specified both `split_fastq_by_size` and `split_fastq_by_part`.' +
+    ' Please specify only one of these parameters. The other should be null.'
+}
+
+// Check that nonsensical combinations of parameters are not set
+if (params.additional_fasta && (params.bwa_index || params.bwamem2_index || params.bowtie_index || params.bowtie2_index)) {
+    exit 1, 'You have specified an additional fasta file and a genome index.' +
+    ' If the genome index is not equivalent to the main fasta file,' +
+    ' then omit the index and allow the pipeline to create it from' +
+    ' the concatenated fasta files.'
+}
+
+if (params.datatype == "mammals" && params.r1_bc_pattern == null){
+    exit 1, 'You have not specified a barcode pattern for mammalian data.' +
+    ' Please specify a barcode pattern using the `r1_bc_pattern` parameter.'
+}
+
```

Comment:
In my opinion any non template validation should go to callingcards.nf. In any case, can you replace `exit 1` with `error()` for better canceling

Reply:
there are two workflows, one more yeast, the other for mammals, so no single `callingcards.nf`. That itself is likely a point to consider -- a single `callingcards.nf` that calls different subworkflows depending on the `params.datatype`